usdt-macro/src/lib.rs

```diff
@@ -31,7 +29,13 @@ use usdt_impl::compile_providers;
 /// dtrace_provider!("test.d");
 /// ```
 ///
-/// One can then instrument the application or library as one might expect:
+/// The macro looks for the file relative to the root of the package, so `"test.d"`
+/// in this case would be in the same directory as `"Cargo.toml"`.
+///
+/// By default probe macros are named `{provider}_{probe}!`. Arguments are passed
+/// via a closure (or function) that returns a tuple. Note that the provided
```

Comment:
Actual functions aren't supported right now. This is one of the sharp-edges of the macro: there's no matcher that identifies both closures and functions, but that also filters out other paths, like `std::collections::Vec`. So it does _have_ to be a closure right now.

Reply:
What wouldn't `expr` work for functions? I tried removing the `tt` pattern and using a function just worked. What cases are we trying to catch here?

Reply:
I could be wrong, but the goal was to allow closures, but disallow things like the actual arguments. Expr does work for functions, but it's overly permissive, as it allows a tuple with the arguments. So I put a token-tree matcher first, to filter out the tuple, which also matches a function.

So one option would be to match on an expression or a path, which matches things like `std::collections::Vec` as well as functions. The path and expr matches would expand to the same branch, and the token-tree would filter out things like tuples. I don't know that any of this needs to happen now, but if/when we want to support functions, this might be one approach.

[Here's a link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=497c610ed59038dd2eeea56c0a6e8b07) to the playground showing this in action, for future reference.

Reply:
Agreed; we can defer doing anything more permissive here.